### PR TITLE
authz: fix external authorization when `maxRequestBytes` is greater than 1MiB

### PR DESCRIPTION
**Pull request by jewertow**
**Please provide a description of this PR:**

External authorization does not work when `envoyExtAuthz{Http,Grpc}.includeRequestBodyInCheck.maxRequestBytes` is greater than 1MiB, because HttpConnectionManager does not buffer more than 1MiB and returns `413 Payload Too Large` if it can't stream the request.

The solution is to add `envoy.filters.http.buffer` before `envoy.filters.http.ext_authz`.

You can quickly reproduce this issue and verify the solution by following steps from this [README](https://github.com/jewertow/istio-playground/tree/master/issues/ext-authz-max-requests-bytes-413-payload-too-large).

### CODE REVIEW

File: pilot/pkg/security/authz/builder/builder.go [hunk 1]
```diff
@@ -303,6 +305,20 @@ func (b Builder) buildHTTP(rules *rbacpb.RBAC, shadowRules *rbacpb.RBAC, provide
...
+	if extauthz.http.GetWithRequestBody().GetMaxRequestBytes() > 1048576 {
+		httpBuffer := &hcm.HttpFilter{
+			Name: wellknown.Buffer,
+			ConfigType: &hcm.HttpFilter_TypedConfig{TypedConfig: protoconv.MessageToAny(&buffer.Buffer{
+				MaxRequestBytes: &wrappers.UInt32Value{Value: extauthz.http.GetWithRequestBody().MaxRequestBytes},
+			})},
+		}
+		// insert buffer between rbac and ext_authz
+		httpFilters = append(httpFilters[:1], httpFilters[0:]...)
+		httpFilters[1] = httpBuffer
```
Comment (howardjohn) [inline pilot/pkg/security/authz/builder/builder.go:319]: nit: can we just build this in-order to begin with instead of shuffling things around?

Comment (jewertow) [reply pilot/pkg/security/authz/builder/builder.go:319]: Do you mean that the buffer should be first?

**Comment by hzxuzhonghu** [comment]
can you set `PerConnectionBufferLimitBytes` on listener

**Comment by jewertow** [comment]
> can you set `PerConnectionBufferLimitBytes` on listener

It would be more complicated, because I would have to check ext_authz filter in the listener builder, and since this issue is rather a rare case, I preferred simpler implementation.

**Comment by kyessenov** [comment]
I think this is an API change. Buffering will break streaming (e.g. xDS). Maybe add an API change to "enable_buffer"? The buffer size probably depends on HTTP version - is it the same on h2?

**Comment by kyessenov** [comment]
> > Buffering will break streaming (e.g. xDS).
> 
> I don't understand it... An XDS connection is initiated by the proxy, so it not handled by an inbound listener that, in addition, does not even exist.

Buffer filter forces request body buffering, and that is incompatible with bi-di requests, like gRPC xDS. The canonical semantics for gRPC is:

```Request → Request-Headers *Length-Prefixed-Message EOS```

The body is left open when the response starts.

> 
> > Maybe add an API change to "enable_buffer"?
> 
> The purpose of this PR is exactly the opposite.

Buffering cannot be enabled  by default. In addition to breaking the streaming requests, it may also add latency and memory overhead - since it has a negative impact on the flow control.

> > The buffer size probably depends on HTTP version - is it the same on h2?
> 
> From my tests, the limit is the same.

What causes overflow in the codec? h2 has per-stream and per-connection limits - it depends how they are configured.

**Comment by jewertow** [comment]
> Buffer filter forces request body buffering, and that is incompatible with bi-di requests, like gRPC xDS. The canonical semantics for gRPC is:
>
> Request → Request-Headers *Length-Prefixed-Message EOS
>
> The body is left open when the response starts.

Sure, I understand it, but how a buffer filter in an HCM in an inbound listener can affect gRPC XDS connection between Envoy and istiod, which is not handled by that listener?

> Buffering cannot be enabled by default. In addition to breaking the streaming requests

Ok, I don't know how http buffer filter work internally, because it's not documented, but streamed requests are buffered anyway - there is a [stream window size](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/core/v3/protocol.proto#envoy-v3-api-field-config-core-v3-http2protocoloptions-initial-stream-window-size) (`This field also acts as a soft limit on the number of bytes Envoy will buffer per-stream in the HTTP/2 codec buffers`).

The listener also must read bytes from a socket with `read()`, so I'm sorry, but I don't understand your position that "buffering cannot be enabled". In my understanding, buffering is performed anyway, and my change adjusts its size. Why is this reasoning wrong?

> What causes overflow in the codec? h2 has per-stream and per-connection limits - it depends how they are configured.

Requests smaller than 1MiB, but greater than the default stream window size (65535 B) worked well, so I conclude that the per-connection limit, which is 1 MiB by default, causes the overflow.

**Comment by kyessenov** [comment]
> > Buffer filter forces request body buffering, and that is incompatible with bi-di requests, like gRPC xDS. The canonical semantics for gRPC is:
> > Request → Request-Headers *Length-Prefixed-Message EOS
> > The body is left open when the response starts.
> 
> Sure, I understand it, but how a buffer filter in an HCM in an inbound listener can affect gRPC XDS connection between Envoy and istiod, which is not handled by that listener?

xDS is given as an example application payload. There are other types of traffic, e.g. CONNECT that has similar issues.

> The listener also must read bytes from a socket with `read()`, so I'm sorry, but I don't understand your position that "buffering cannot be enabled". In my understanding, buffering is performed anyway, and my change adjusts its size. Why is this reasoning wrong?

Flow control is dynamic. Under load, the windows shrink and the proxy needs to push back against buffering. With a fixed buffer window, the pushback is delayed, and the proxy runs a risk of memory overflow.

> Requests smaller than 1MiB, but greater than the default stream window size (65535 B) worked well, so I conclude that the per-connection limit, which is 1 MiB by default, causes the overflow.

The defaults in both Istio and Envoy are not production safe. Istio advertises a very large stream window AFAIK which can cause an OOM if the peer is not picking something smaller.

**Comment by jewertow** [comment]
> Maybe add an API change to "enable_buffer"?

Oh, do you mean extending `includeRequestBodyInCheck`? I would like it, because the buffer may overflow while receiving many files simultaneously.

**Comment by yuval-k** [comment]
> @kyessenov I think that all what we talk about is irrelevant for this change, because buffering is always enabled when ext_authz has `with_request_body`, and my change only adjust size of the buffer, because HCM does not inherit buffer settings from ext_authz. It will not change anything in request processing.

Your change adds buffering to all routes under the listener. while ext-authz might be enabled for only some of them.
so if a user only has 1 route that he wants to use with ext-authz, now all of a sudden other unrelated routes are subject to buffering too and may break.

**Comment by jewertow** [comment]
> Your change adds buffering to all routes under the listener.

To be precise, it applies to all routes under a filter.

> while ext-authz might be enabled for only some of them.

There is also [BufferPerRoute](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/buffer/v3/buffer.proto#extensions-filters-http-buffer-v3-bufferperroute). I will try it.

> unrelated routes are subject to buffering too and may break.

I don't like the term "break" in this context, because as I said in the previous comment, buffering is always enabled when `includeRequestBodyInCheck` is enabled, and it does not break anything, including gRPC.

**Comment by yuval-k** [comment]
> To be precise, it applies to all routes under a filter.

You added an http filter to the HCM. this applies to all virtual-hosts and routes in the HCM's route configuration.

> I don't like the term "break" in this context, because as I said in the previous comment, buffering is always enabled when includeRequestBodyInCheck is enabled, and it does not break anything, including gRPC.

This would be true if buffering would only happen if extauth is happening. The key point is the one above. you can have routes that will not have ext-auth under this HCM. Taking [this exmaple](https://istio.io/latest/docs/tasks/security/authorization/authz-custom/#enable-with-external-authorization) from the docs: 

```yaml
apiVersion: security.istio.io/v1
kind: AuthorizationPolicy
...
  rules:
  # The rules specify when to trigger the external authorizer.
  - to:
    - operation:
        paths: ["/headers"]
```

if the user goes to "/not-headers" ext-auth won't be active, but buffering will be active.

**Comment by kyessenov** [comment]
In general, I think ext_authz with body buffering might be flawed architecturally. It puts too much burden to buffer potentially malicious requests in memory with a significant buffer (>1MB) which permits a denial of service attack. Can you explain what exactly you're doing with 1MB from the start? A streaming version with ext_proc might work better.

CC @rshriram

**Comment by sschepens** [comment]
Would we maybe want to support `ext_proc` instead of doing buffered `ext_authz` ? `ext_proc` supports streaming the body to the external service without buffering. https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_proc/v3/processing_mode.proto#envoy-v3-api-enum-extensions-filters-http-ext-proc-v3-processingmode-bodysendmode

**Comment by sschepens** [comment]
On another note, `ext_authz` [doc](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto#envoy-v3-api-msg-extensions-filters-http-ext-authz-v3-buffersettings) suggests that `ext_authz` is doing the buffering and that adding `buffer` filter should not be necessary.

@jewertow are you sure that `buffer` filter must be added? documentation also suggests that if body surpasses the maximum defined in the `ext_authz` settings, the a 413 will be returned, unless `allow_partial_message` is `true`.

**Comment by jewertow** [comment]
> ext_authz definitely seems to be doing [buffering by itself](https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/ext_authz/ext_authz.cc#L179-L193) or directly [setting the decoder buffer limit](https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/ext_authz/ext_authz.cc#L165-L167) if partial is disallowed

Oh right... `with_request_body` always enables buffering, but when `allow_partial_message: true`, then the ext_authz ignores `max_request_bytes`, so the buffer limit is 1MiB, and that's why 413 is returned for requests greater than 1MiB.

> Can you explain what exactly you're doing with 1MB from the start? A streaming version with ext_proc might work better.

@kyessenov Ok, the use case is authorization of a request based on HMAC of an uploaded file. This is exactly the same use case like the original PR introducing `with_request_body` to the ext_authz filter. But as you can see above, we don't need another buffer filter, because the problem was caused by `allow_partial_message: true`, so I'm closing this PR.

**Pull request closed without merging**